core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -203,22 +204,25 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
      *
      * @param unassignedIterator iterator over unassigned shards
      * @param shard the shard which might be delayed
+     * @return true iff there was a change to the unassigned info
      */
-    public void ignoreUnassignedIfDelayed(RoutingNodes.UnassignedShards.UnassignedIterator unassignedIterator, ShardRouting shard) {
+    public boolean ignoreUnassignedIfDelayed(RoutingNodes.UnassignedShards.UnassignedIterator unassignedIterator, ShardRouting shard) {
         if (shard.unassignedInfo().isDelayed()) {
             logger.debug("{}: allocation of [{}] is delayed", shard.shardId(), shard);
             /**
              * mark it as changed, since we want to kick a publishing to schedule future allocation,
              * see {@link org.elasticsearch.cluster.routing.RoutingService#clusterChanged(ClusterChangedEvent)}).
              */
-            unassignedIterator.removeAndIgnore();
+            return unassignedIterator.removeAndIgnore(Optional.empty());
         }
+        return false;
     }
 
     /**
      * Can the shard be allocated on at least one node based on the allocation deciders.
      */
-    private boolean canBeAllocatedToAtLeastOneNode(ShardRouting shard, RoutingAllocation allocation) {
+    private Decision canBeAllocatedToAtLeastOneNode(ShardRouting shard, RoutingAllocation allocation) {
```

Comment:
lets not make this method return null. Have it return Decision.NO instead. Also add some javadocs what the result means.